src/analysis/processing/qgsalgorithmexplodehstore.cpp

```diff
@@ -0,0 +1,184 @@
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Exploded" ) ) );
+}
+
+QVariantMap QgsExplodeHstoreAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+  int attrSourceCount = source->fields().count();
+
+  QString fieldName = parameterAsString( parameters, QStringLiteral( "FIELD" ), context );
+  int fieldIndex = source->fields().lookupField( fieldName );
+  if ( fieldIndex < 0 )
+    throw QgsProcessingException( QObject::tr( "Invalid HStore field" ) );
+
+  QStringList expectedFields;
+  QString fieldList = parameterAsString( parameters, QStringLiteral( "EXPECTED_FIELDS" ), context );
+  if ( ! fieldList.trimmed().isEmpty() )
+  {
+    expectedFields = fieldList.split( ',' );
+  }
+
+  QList<QString> fieldsToAdd;
+  QHash<QgsFeatureId, QVariantMap> hstoreFeatures;
+  QList<QgsFeature> features;
+
+  double step = source->featureCount() > 0 ? 50.0 / source->featureCount() : 1;
+  int i = 0;
+  QgsFeatureIterator featIterator = source->getFeatures( );
+  QgsFeature feat;
+  while ( featIterator.nextFeature( feat ) )
+  {
+    i++;
+    if ( feedback->isCanceled() )
+      break;
+
+    double progress = i * step;
+    if ( progress >= 50 )
+      feedback->setProgress( 50.0 );
+    else
+      feedback->setProgress( progress );
+
+    QVariantMap currentHStore = QgsHstoreUtils::parse( feat.attribute( fieldName ).toString() );
+    for ( const QString &key : currentHStore.keys() )
+    {
+      if ( expectedFields.isEmpty() && ! fieldsToAdd.contains( key ) )
+        fieldsToAdd.insert( 0, key );
+    }
+    hstoreFeatures.insert( feat.id(), currentHStore );
+    features.append( feat );
+  }
+
+  if ( ! expectedFields.isEmpty() )
+  {
+    fieldsToAdd = expectedFields;
+  }
+
+  QgsFields hstoreFields;
+  for ( const QString &fieldName : fieldsToAdd )
+  {
+    hstoreFields.append( QgsField( fieldName, QVariant::String ) );
+  }
+
+  QgsFields outFields = QgsProcessingUtils::combineFields( source->fields(), hstoreFields );
+
+  QString sinkId;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, sinkId, outFields, source->wkbType(), source->sourceCrs() ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  QList<int> fieldIndicesInput = QgsProcessingUtils::fieldNamesToIndices( QStringList(), source->fields() );
+  int attrCount = attrSourceCount + fieldsToAdd.count();
+  QgsFeature outFeature;
+  step = 50.0 / i;
```

Comment:
```suggestion
  step = !features.empty() ? 50.0 / features.count() : 1;
```

Reply:
Are you sure? At that stage, `i` contains the exact number of features in the source layer. So it's better to use it to update the progress bar instead of `1` which will result in a uncorrect progress bar when the feature count is unknown.

Reply:
It took me a while to understand the logic with `i`, but the features list here is also guaranteed to have the exact feature count and I think this makes it slightly more readable/future proof.

Reply:
Ah, it was before my switch to a QList container. It was wrong when I was using the source layer because of unknown featureCount. Sorry